File: DP_execrise/GoUpstairs.java
package niuke;

//台阶问题练习题
public class GoUpstairs {
	public int countWays(int n) {
		int dp[] = new int[n + 1];// dp[i] 表示上i个台阶的方法数

		dp[1] = 1;// 初始化dp[1]上一个台阶就一种方法
		dp[2] = 2;// 上二个台阶就2种 一次上一个台阶或者一次上2个台阶

		for (int i = 3; i <= n; i++) {
			dp[i] = (dp[i - 1] + dp[i - 2]) % 1000000007;
		}

		return dp[n];
	}
}

File: 第十二章_动态规划/台阶问题练习题/GoUpstairs.cpp
#include <iostream>
using namespace std;

const int N=100000;
const int MOD=1000000007;


//dp[i]=k 代表上i个台阶的方法数为k
int dp[N+1];

class GoUpstairs {
public:
    int countWays(int n) {
        dp[1]=1;//上一个台阶只有一种方法
		dp[2]=2;//上两个台阶有两种，一次上一级或者一次上两级

		for(int i=3;i<=n;i++){
			dp[i]=(dp[i-1]+dp[i-2])%MOD;
			/*那么其实上第i级台阶可以依赖于两种情况得到:
			  1.一种就是上了i-1级之后再上一级
			  2.一种就是上了i-2级之后再上二级

			  两种相加即可得到
			*/
		}

		return dp[n];
    }
};